File: src/main/java/cu/suitetecsa/sdk/nauta/scraper/FormParserImpl.java
package cu.suitetecsa.sdk.nauta.scraper;

import cu.suitetecsa.sdk.nauta.network.HttpResponse;
import org.jetbrains.annotations.NotNull;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.util.AbstractMap.SimpleEntry;
import java.util.HashMap;
import java.util.Map;

public class JsoupFormParser implements FormParser {

    private Document getHtml(@NotNull HttpResponse httpResponse) {
        return Jsoup.parse(httpResponse.getText());
    }
    @Override
    public SimpleEntry<String, Map<String, String>> parseActionForm(@NotNull HttpResponse httpResponse) {
        return parseForm(getHtml(httpResponse).selectFirst("form[action]"));
    }

    @Override
    public SimpleEntry<String, Map<String, String>> parseLoginForm(@NotNull HttpResponse httpResponse) {
        return parseForm(getHtml(httpResponse).selectFirst("form.form"));
    }

    SimpleEntry<String, Map<String, String>> parseForm(Element form) {
        if (form != null) {
            HashMap<String, String> data = getInputs(form);
            String url = form.attr("action");
            return new SimpleEntry<>(url, data);
        }
        return null;
    }

    HashMap<String, String> getInputs(@NotNull Element form) {
        HashMap<String, String> inputs = new HashMap<>();
        for (Element input : form.select("input[name]")) {
            inputs.put(input.attr("name"), input.attr("value"));
        }
        return inputs;
    }
}

File: src/main/kotlin/io/github/suitetecsa/sdk/nauta/scraper/FormParserImpl.kt
package io.github.suitetecsa.sdk.nauta.scraper

import io.github.suitetecsa.sdk.nauta.network.HttpResponse
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import java.util.*

internal class FormParserImpl : FormParser {
    private fun getHtml(httpResponse: HttpResponse): Document {
        return Jsoup.parse(httpResponse.text)
    }

    override fun parseActionForm(httpResponse: HttpResponse): AbstractMap.SimpleEntry<String, Map<String, String>> {
        return parseForm(getHtml(httpResponse).selectFirst("form[action]"))!!
    }

    override fun parseLoginForm(httpResponse: HttpResponse): AbstractMap.SimpleEntry<String, Map<String, String>> {
        return parseForm(getHtml(httpResponse).selectFirst("form.form"))!!
    }

    private fun parseForm(form: Element?): AbstractMap.SimpleEntry<String, Map<String, String>>? {
        if (form != null) {
            val data = getInputs(form)
            val url = form.attr("action")
            return AbstractMap.SimpleEntry(url, data)
        }
        return null
    }

    private fun getInputs(form: Element): HashMap<String, String> {
        val inputs = HashMap<String, String>()
        for (input in form.select("input[name]")) {
            inputs[input.attr("name")] = input.attr("value")
        }
        return inputs
    }
}
